precompiles/bank/bank.go

```diff
@@ -0,0 +1,208 @@
+	bankKeeper bank.Keeper,
+	fungibleKeeper fungiblekeeper.Keeper,
+	cdc codec.Codec,
+	kvGasConfig storetypes.GasConfig,
+) *Contract {
+	// Instantiate the ZRC20 ABI only one time.
+	// This avoids instantiating it every time deposit or withdraw are called.
+	zrc20ABI, err := zrc20.ZRC20MetaData.GetAbi()
+	if err != nil {
+		return nil
+	}
+
+	return &Contract{
+		BaseContract:   ptypes.NewBaseContract(ContractAddress),
+		bankKeeper:     bankKeeper,
+		fungibleKeeper: fungibleKeeper,
+		zrc20ABI:       zrc20ABI,
+		cdc:            cdc,
+		kvGasConfig:    kvGasConfig,
+	}
+}
+
+// Address() is required to implement the PrecompiledContract interface.
+func (c *Contract) Address() common.Address {
+	return ContractAddress
+}
+
+// Abi() is required to implement the PrecompiledContract interface.
+func (c *Contract) Abi() abi.ABI {
+	return ABI
+}
+
+// RequiredGas is required to implement the PrecompiledContract interface.
+// The gas has to be calculated deterministically based on the input.
+func (c *Contract) RequiredGas(input []byte) uint64 {
+	// get methodID (first 4 bytes)
+	var methodID [4]byte
+	copy(methodID[:], input[:4])
+	// base cost to prevent large input size
+	baseCost := uint64(len(input)) * c.kvGasConfig.WriteCostPerByte
+	if ViewMethod[methodID] {
+		baseCost = uint64(len(input)) * c.kvGasConfig.ReadCostPerByte
+	}
+
+	if requiredGas, ok := GasRequiredByMethod[methodID]; ok {
+		return requiredGas + baseCost
+	}
+
+	// Can not happen, but return 0 if the method is not found.
+	return 0
+}
+
+// Run is the entrypoint of the precompiled contract, it switches over the input method,
+// and execute them accordingly.
+func (c *Contract) Run(evm *vm.EVM, contract *vm.Contract, readOnly bool) ([]byte, error) {
+	method, err := ABI.MethodById(contract.Input[:4])
+	if err != nil {
+		return nil, err
+	}
+
+	args, err := method.Inputs.Unpack(contract.Input[4:])
+	if err != nil {
+		return nil, err
+	}
+
+	stateDB := evm.StateDB.(ptypes.ExtStateDB)
+
+	switch method.Name {
+	// Deposit and Withdraw methods are both not allowed in read-only mode.
+	case DepositMethodName, WithdrawMethodName:
+		if readOnly {
+			return nil, ptypes.ErrUnexpected{
+				Got: "method not allowed in read-only mode " + method.Name,
+			}
+		}
+
+		var res []byte
+		execErr := stateDB.ExecuteNativeAction(contract.Address(), nil, func(ctx sdk.Context) error {
+			if method.Name == DepositMethodName {
+				fmt.Println("DEBUG: bank.Run(): DepositMethodName: ExecuteNativeAction c.deposit()")
+				res, err = c.deposit(ctx, evm, contract, method, args)
+			} else {
+				fmt.Println("DEBUG: bank.Run(): WithdrawMethodName: ExecuteNativeAction c.withdraw()")
+				res, err = c.withdraw(ctx, evm, contract, method, args)
+			}
+			return err
+		})
+		if execErr != nil {
+			fmt.Printf("DEBUG: bank.Run(): execErr %s", execErr.Error())
+			return nil, err
+		}
+		return res, nil
+
+	case BalanceOfMethodName:
+		fmt.Println("DEBUG: bank.Run(): BalanceOfMethodName")
```

Comment:
Would these print lines be removed before the final merge ? same for the lines in this file

Reply:
Those are to be removed after setting up all the unit and e2e testing.